MetaMask/metamask-extension
# refactor: Refactor extension storage classes

## **Description**
adds a new class that models the shared surface of the two storage solutions we utilize, and then creates two new classes designed to extend from that one. The goal of this Pull request is to consolidate some of the operations of state management that were previously being implemented in background.js (the process of creating a new state tree). We relied upon a bullish value being returned from the `.get()` method to know to generate a new first time state object fro the user. This logic has been moved into the storage class so that determining whether to use a new state tree is handled within the storage system itself. 

This is a step towards providing a fail safe for state corruption. The new class will be utilized for this additional functionality.

## **Related issues**

Fixes:

## **Manual testing steps**
No functional changes should have occurred. Manual regression cases to test include:
- make some changes such as a adding a network or changing settings. Close the browser. Re-open the browser and open metamask. The data should be unchanged.
- Install an old version... such as 11.10.0. Make some changes. Upgrade to the build generated from this branch. Verify that the state of the wallet, and its data, is as expected.

## **Screenshots/Recordings**



### **Before**



### **After**



## **Pre-merge author checklist**


## **Pre-merge reviewer checklist**
